controller/src/lib.rs

```diff
@@ -20,114 +25,204 @@ pub mod plan;
+            Err(SchedulerError::NoDroneAvailable) => return Ok(ScheduleResponse::NoDroneAvailable),
+        };
+        match spawn_backend(nats.clone(), drone, &sr.clone()).await? {
+            (res, Some(st)) => {
+                match state.state().get_listener(st) {
+                    Ok(listener) => {
+                        *l = Some(listener);
                     }
-                    Err(error) => {
-                        tracing::warn!(?error, "Scheduler returned error.");
-                        ScheduleResponse::NoDroneAvailable
+                    Err(SequenceNumberInThePast) => {
+                        tracing::warn!("tried to insert notifier after valid time");
                     }
-                }
+                };
+                Ok(res)
             }
-            Err(error) => match error {
-                SchedulerError::NoDroneAvailable => {
-                    tracing::warn!("No drone available.");
-                    ScheduleResponse::NoDroneAvailable
-                }
-            },
-        };
+            (res, None) => Ok(res),
+        }
+    } else {
+        match scheduler.schedule(&cluster_name, Utc::now()) {
+            Ok(drone_id) => Ok(spawn_backend(nats.clone(), drone_id, &sr.clone()).await?.0),
+            Err(SchedulerError::NoDroneAvailable) => Ok(ScheduleResponse::NoDroneAvailable),
+        }
+    }
+}
+
+type WaitMap = HashMap<String, Arc<Mutex<Option<ClosableNotify>>>>;
+pub async fn run_scheduler(nats: TypedNats, state: StateHandle) -> NeverResult {
+    let scheduler = Scheduler::new(state.clone());
+    let mut schedule_request_sub = nats.subscribe(ScheduleRequest::subscribe_subject()).await?;
+    tracing::info!("Subscribed to spawn requests.");
+    let mut lock_to_ready: WaitMap = HashMap::new();
+
+    while let Some(schedule_request) = schedule_request_sub.next().await {
+        tracing::info!(metadata=?schedule_request.value.metadata.clone(), "Got spawn request");
 
-        schedule_request.respond(&result).await?;
+        let nats = nats.clone();
+        let schedule_request = schedule_request.clone();
+        let state = state.clone();
+        let lock = schedule_request.value.lock.clone();
+        let lock_to_ready = &mut lock_to_ready;
+        let lock_mut_maybe = lock.clone().map(|lock| {
+            lock_to_ready
+                .entry(lock)
+                .or_insert(Arc::new(Mutex::new(None)))
+                .clone()
+        });
+        let scheduler = scheduler.clone();
+        tokio::spawn(async move {
+            let sr = schedule_request.value.clone();
+            tracing::info!(?lock, "scheduling lock");
+            let cluster_name = sr.cluster.clone();
+
+            let Ok(response) = dispatch(
+                state.clone(),
+                cluster_name,
+                sr.clone(),
+                scheduler.clone(),
+                nats.clone(),
+                sr.lock.clone().zip(lock_mut_maybe)
```

Comment:
I'm not a fan of zips on options. Probably cleanest to create a named variable from this.